cranelift/codegen/src/isa/x86/enc_tables.rs

```diff
@@ -1219,6 +1219,90 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_block(pos.func, done);
 }
 
+// Lanes of an I32x4 filled with the max signed integer values converted to an F32x4. TODO flip?
+static MAX_SIGNED_I32X4S_AS_F32X4S: [u8; 16] = [
+    0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00,
+];
+
+/// This legalization converts a vector of 32-bit floating point lanes to unsigned integer lanes
+/// using VCVTPS2UDQ if AVX512 is available or to a longer sequence of NaN quieting and lane
+/// otherwise. This logic is separate from [expand_fcvt_to_uint_sat] above (the scalar version),
+/// only due to how the transform groups are set up; TODO if we change he SIMD legalization groups,
+/// then this logic could be merged into [expand_fcvt_to_uint_sat] (see
+/// https://github.com/bytecodealliance/wasmtime/issues/1745).
+fn expand_fcvt_to_uint_sat_vector(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Unary {
+        opcode: ir::Opcode::FcvtToUintSat,
+        arg,
+    } = pos.func.dfg[inst]
+    {
+        let controlling_type = pos.func.dfg.ctrl_typevar(inst);
+        if controlling_type == I32X4 {
+            debug_assert_eq!(pos.func.dfg.value_type(arg), F32X4);
+            let x86_isa = isa
+                .as_any()
+                .downcast_ref::<isa::x86::Isa>()
+                .expect("the target ISA must be x86 at this point");
+            if x86_isa.isa_flags.use_avx512vl_simd() || x86_isa.isa_flags.use_avx512f_simd() {
```

Comment:
Are the rounding semantics of the two alternative implementations equivalent?  For the short case, `vcvtps2udq`, the Intel manual says "When a conversion is inexact, the value returned is rounded according to the rounding control bits in the MXCSR register or the embedded rounding control bits."  The slow path appears to use `cvtt2si` (presumably, really, `cvttps2pi`), and all the `tt` variant conversion insns are round-towards-zero: "When a conversion is inexact, a truncated (round toward zero) result is returned."

Reply:
This is a good point: I had hoped MXCSR would have the right rounding mode but it does not. I added a `run` filetest that passes with the long sequence and fails with the AVX512 sequence: it looks like the machines running the CI have AVX512 and therefore fail. I started playing around with a new recipe that embeds the rounding control (e.g. `EvexContext::RoundingRegToRegFP { rc: EvexRoundingControl::RZ }` but this didn't immediately fix the test case (the emitted bytes aren't exactly right) so I will have to play around with this a bit more.